compiler/src/iree/compiler/Codegen/ExternalInterfaces/Utils.cpp

```diff
@@ -23,6 +24,73 @@
 namespace mlir::iree_compiler::IREE {
 using IREE::Codegen::MaterializeEncodingInfo;
 
+void adjustTileSizesForBitcast(RankedTensorType type,
+                               MaterializeEncodingInfo &info) {
+  auto encoding =
+      dyn_cast_if_present<Encoding::EncodingAttr>(type.getEncoding());
+  if (!encoding) {
+    return;
+  }
+
+  // Only adjust if the encoding has an original_element_type, indicating a
+  // bitcast occurred.
+  auto originalElementTypeAttr = encoding.getOriginalElementType();
+  if (!originalElementTypeAttr) {
+    return;
+  }
+
+  Type originalType = originalElementTypeAttr.getValue();
+  Type storageType = type.getElementType();
+
+  // No adjustment needed if types are the same.
+  if (originalType == storageType) {
+    return;
+  }
+
+  assert(originalType.isIntOrFloat() && storageType.isIntOrFloat() &&
+         "expected int or float element types in encoding");
+  unsigned originalBits = originalType.getIntOrFloatBitWidth();
+  unsigned storageBits = storageType.getIntOrFloatBitWidth();
+
+  // One bit width must be a multiple of the other.
+  assert((storageBits >= originalBits ? storageBits % originalBits
+                                      : originalBits % storageBits) == 0 &&
+         "bit widths must be multiples of each other");
+
+  // Adjust the innermost tile size based on the bit width ratio between
+  // original and storage types. The innermost dimension is where packing
+  // occurs.
+  if (info.innerTileSizes.empty()) {
+    return;
+  }
+  int64_t &innermostTileSize = info.innerTileSizes.back();
+  if (innermostTileSize == ShapedType::kDynamic) {
+    return;
+  }
```

Comment:
I don't find the link right now, but w prefer using `ShapedType::isDynamic(innermostTileSize)`.